### PR TITLE
Add temp directory (discussion windows  bug fixes)

### PR DESCRIPTION
**Pull request by LattaruoloAndrea**
Hello I have been working to check what was not working on windows, with the changes made on download.py now it should work. I also added a parameter to set the temp directory as well, if it's not needed we can remove it #150  should be fixed with this.

### CODE REVIEW

File: src/cdsetool/download.py [hunk 1]
```diff
@@ -38,6 +38,7 @@ def download_feature(
     log = _get_logger(options)
     url = _get_feature_url(feature)
     title = feature.get("properties").get("title")
+    temp_dir = _get_temp_dir(options)
```
Comment (pjonsson) [inline src/cdsetool/download.py:41]: I think `temp_dir = options.get("tmpdir") or None` will give the result you want, and that should also give the nice behavior when the user specifies `"tmpdir": ""` in options.

**Comment by pjonsson** [comment]
This looks like it implements my suggestion in #51 to be able to specify where to put the temporary directory (nice!), accidentally reverts my empty log-message fix, and changes something with the logic of temp file closing.

I think being allowed to specify the temporary directory location isn't controversial, could you please put that in a separate PR against latest main?

I don't understand the logic changes for file closing, but #164 fixed a bug where files weren't closed properly. Can you check if the logic change from this PR is still required on top of latest main?

File: src/cdsetool/download.py (hunk 1 above)
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py:41]: I tried to use the same logic as for log and url that it was created a function specifically but inside the function there is the logic that you mentioned.

**Comment by LattaruoloAndrea** [comment]
I will try to create a different PR for the temp file, with the current implementations I found that there are a few things that I don't like.
```python
                tmp_file = os.path.join(temp_dir, "download.zip")
                with open(tmp_file, "wb") as file:
                    # Server might not send all bytes specified by the
                    # Content-Length header before closing connection.
                    # Log as a warning and try again.
                    try:
                        for chunk in response.iter_content(chunk_size=1024 * 1024 * 5):
                            file.write(chunk)
                            status.add_progress(len(chunk))
                    except (
                        ChunkedEncodingError,
                        ConnectionResetError,
                        ProtocolError,
                    ) as e:
                        log.warning(e)
                        continue
                # Close file before copy so all buffers are flushed.
                shutil.copy(tmp_file, result_path)
                return filename
```
Is it better to change the  `tmp_file = os.path.join(temp_dir, "download.zip")` with `tmp_file = tempfile.NamedTemporaryFile(dir=temp_dir, prefix=filename.replace(".zip", "")).name` since your implementations looks like that all the temp file will be having the same nave "download.zip" in the temp dir which in concurrent threads I guess that's an error, while `tempfile.NamedTemporaryFile()` creates a random Id and we add a prefix with the name of the file

**Comment by pjonsson** [comment]
> Is it better to change the `tmp_file = os.path.join(temp_dir, "download.zip")` with `tmp_file = tempfile.NamedTemporaryFile(dir=temp_dir, prefix=filename.replace(".zip", "")).name` since your implementations looks like that all the temp file will be having the same name "download.zip" in the temp dir which in concurrent threads I guess that's an error, while `tempfile.NamedTemporaryFile()` creates a random Id and we add a prefix with the name of the file.

You get a unique temporary directory from each call to `TemporaryDirectory()`, so both main branch and this PR should be thread-safe.

> And then in the current implementations there is no logic to delete the tmp_file after it was copyied if I am not wrong. I guess if you manage to work on these I can try again on windows and check if it works.

The temporary directory and everything inside it is removed when the program leaves the with-statement that has the 'with tempfile.TemporaryDirectory() as ..:`.

**Comment by LattaruoloAndrea** [comment]
>You get a unique temporary directory from each call to `TemporaryDirectory()`, so both main branch and this PR should be thread-safe.

What if you specify the dir in TemporaryDirectory() will it creates different directories inside the dir specified?

> The temporary directory and everything inside it is removed when the program leaves the with-statement that has the `with tempfile.TemporaryDirectory() as ..:`.
Great! So I can test it on windows basically

**Comment by pjonsson** [comment]
> > You get a unique temporary directory from each call to `TemporaryDirectory()`, so both main branch and this PR should be thread-safe.
> 
> What if you specify the dir in TemporaryDirectory() will it creates different directories inside the dir specified?

Yes!

**Review by pjonsson — APPROVED** [review]

**Comment by LattaruoloAndrea** [comment]
I just added the prefix with filename inside the tempDir so it get's easier to track what it is being downloaded. I guess it is better to leave it in the dirTemp instead of the file name.

**Review by subnet001 — APPROVED** [review]

**Pull request merged**